include/utf8.h

```diff
@@ -0,0 +1,86 @@
+#ifndef UTF8_CPP_H
+#define UTF8_CPP_H
+
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace Url
+{
+
+    /**
+     * Work between unicode code points and their UTF-8-encoded representation.
+     */
+    struct Utf8
+    {
+        /**
+         * The type we use to represent Unicode codepoints.
+         */
+        typedef long codepoint_t;
```

Comment:
I'm really used to using `uint32_t`, but I'd like to learn something here -- I have this recollection of being discouraged from using `*int32_t` when I want the guarantee that the type holds _at least_ 32 bits and to only use that type when I need _exactly_ 32 bits. I seem to recall that the motivation is that the type is chosen to be whatever's fastest on the host architecture while only providing the guarantee that it holds at least 32 bits. Does that ring a bell? It was a sort of half-remembered thing.

At the very least I'm happy to make it `unsigned`, and if we prefer it to be `uint32_t` then we could do that, too.